sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -158,49 +151,67 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
+                                            .getThresholdStep().multipliedBy(monoList.size() - 1))));
+                            }
+                        }
+                    });
+            }
         }
 
-        for (URI endpoint : speculativeProcessor.getRegionsToSpeculate(config, this.transportClient
-            .getGlobalEndpointManager().getReadEndpoints())) {
-            if (request.requestContext.locationEndpointToRoute != endpoint) {
-                RxDocumentServiceRequest newRequest = request.clone();
-                newRequest.requestContext.routeToLocation(endpoint);
-                requests.add(newRequest);
-                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout)
-                    .delaySubscription(speculativeProcessor.getThreshold(config).plus(speculativeProcessor.getThresholdStepDuration(config, monoList.size() - 1))));
-            }
+        // If the above conditions are not met, then we will just return the original request
+        if (monoList.isEmpty()) {
+            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
         }
 
-        return Mono.firstWithValue(monoList).map(storeResponse ->
-            {
-                for (RxDocumentServiceRequest r : requests) {
-                    CosmosDiagnostics diagnostics = r.requestContext.cosmosDiagnostics;
-                    // for the successful request region, we will use the actual latency
-                    if (r.getActivityId().toString().equals(storeResponse.getActivityId())) {
-                        speculativeProcessor.onResponseReceived(r.requestContext.locationEndpointToRoute,
-                            diagnostics.getDuration());
-                    }
-                }
-                return storeResponse;
-            });
+        return Mono.firstWithValue(monoList);
+    }
+
+    private List<URI> getApplicableEndPoints(RxDocumentServiceRequest request) {
+        if (request.isReadOnlyRequest()) {
+            return this.transportClient.getGlobalEndpointManager().getApplicableReadEndpoints(request);
+        } else if (request.getOperationType().isWriteOperation()) {
+            return this.transportClient.getGlobalEndpointManager().getApplicableWriteEndpoints(request);
+        }
+        return null;
     }
 
     private boolean shouldSpeculate(RxDocumentServiceRequest request) {
-        if (speculativeProcessor == null) {
+        if (request.requestContext.getEndToEndOperationLatencyPolicyConfig() == null) {
+            return false;
+        }
+        if (request.getResourceType() != ResourceType.Document) {
             return false;
         }
-        if (!request.isReadOnlyRequest() && request.getResourceType() != ResourceType.Document) {
+
+        if (!request.isReadOnlyRequest()
+            || (request.getOperationType().isWriteOperation() && !request.getNonIdempotentWriteRetriesEnabled())) {
```

Comment:
it will always return false for write operation. I think we should remove `!request.isReadOnlyRequest() `?